Added add_evidence custom function

kelby-shelton opened the pull request:
Added add_evidence custom function based on Phantom REST API: https://docs.splunk.com/Documentation/Phantom/4.10/PlatformAPI/RESTEvidence

kelby-shelton committed:
Added add_evidence custom function

Added add_evidence custom function based on Phantom REST API: https://docs.splunk.com/Documentation/Phantom/4.10/PlatformAPI/RESTEvidence

kelby-shelton committed:
Update add_evidence.py

Changed response['success'] to response.get('success') to account for error condition on already added evidence.

philroyer-phantom commented on custom_functions/add_evidence.py:
```diff
@@ -0,0 +1,40 @@
+def add_evidence(container_id=None, object_id=None, content_type=None, **kwargs):
```
I think in order to stay consistent with the UI we should change this custom function name to "mark_evidence".

philroyer-phantom commented on custom_functions/add_evidence.py:
```diff
@@ -0,0 +1,40 @@
+    Custom function implementation of REST API for Evidence
+    
+    Args:
+        container_id (CEF type: phantom container id): Id of the container to which you are adding evidence.
```
I think we should call this input just "container", and use logic like the following to accept either a container object or a container id:
```
     # handle the container input, which can be a container dictionary or a container id as an integer
     if isinstance(container, dict) and container['id']:
         container_id = container['id']
     elif isinstance(container, int):
         container_id = container
     else:
         raise TypeError("The input 'container' is neither a container dictionary nor an int, so it cannot be used")

philroyer-phantom commented on custom_functions/add_evidence.py:
```diff
@@ -0,0 +1,40 @@
+def add_evidence(container_id=None, object_id=None, content_type=None, **kwargs):
+    """
+    Custom function implementation of REST API for Evidence
```
Just to make this a little more readable, I would instead use this description:
`Mark an object as Evidence in a container.`

philroyer-phantom commented on custom_functions/add_evidence.py:
```diff
@@ -0,0 +1,40 @@
+    Args:
+        container_id (CEF type: phantom container id): Id of the container to which you are adding evidence.
+        object_id (CEF type: *): Id of object to be added - artifact id, note id, etc.
+        content_type (CEF type: *): The content type of the object to add as evidence. One of the types:
```
For readability I would change this description to the following:

```
The content type of the object to add as evidence which must be one of the following:
            
            containerattachment
            artifact
            actionrun
            container
            note

philroyer-phantom commented on custom_functions/add_evidence.py:
```diff
@@ -0,0 +1,40 @@
+    # Post data to evidence endpoint
+    response = phantom.requests.post(uri=url, json=data, verify=False).json()
+    # If successful add evidence id to outputs
+    if response.get('success') == 'true':
```
First of all, I think this check is incorrect, because the success value should be True, and True does not equal 'true'.

Second, if you run this against an invalid content object or content type, there is currently no error. I think we should check if the response contains `'failed': True` as well. If it does not contain 'success' or `failed', then it is also no good.

philroyer-phantom reviewed (changes requested):
Overall it looks good, but I had some feedback. The biggest are the custom function name change and checking for API failures.

kelby-shelton committed:
made recommended changes

changed custom function name, added additional error checking, and added support for container objects

kelby-shelton reviewed (commented):
Made requested changes in new commit: https://github.com/phantomcyber/playbooks/pull/15/commits/6f089ee967c2f8452bd4092f89e676553759c1dc

philroyer-phantom commented on custom_functions/mark_evidence.py:
```diff
@@ -0,0 +1,66 @@
+        object_id (CEF type: *): Id of object to be added - artifact id, note id, etc.
+        content_type (CEF type: *): The content type of the object to add as evidence which must be one of the following:
+                        
+                        containerattachment
```
Sorry I missed this the first time, but the containerattachment type seems to be difficult to use. I think we should add a new type called "vault_id" which should accept the sha1 hash of the file. That is the standard way to pass around references to a vault file. Then we should use code like the following to get the id:

```python
    vault_id = '8fe50491d77eace578cfd2862a14c93dc902be43'
    success, message, info = phantom.vault_info(
        vault_id=vault_id
    )

    phantom.debug(
        'phantom.vault_info results: success: {}, message: {}, info: {}'\
        .format(success, message, info)
    )
    
    container_attachment_id = info[0]['id']
    phantom.debug(container_attachment_id)

kelby-shelton replied:
Had to make input content_type = vault_id and then translate in the code since VPE was having a problem with "containerattachment"

philroyer-phantom commented on custom_functions/mark_evidence.py:
```diff
@@ -0,0 +1,66 @@
+    if response.get('success'):
+        outputs['id'] = response['id']
+    elif response.get('failed') and response.get('message'):
+        phantom.debug(f"{content_type} \'{container_id}\' {response['message']}")
```
This should raise a RuntimeError to make the CF fail.

kelby-shelton replied:
Adjusted to only put the debug if the message == "Already added as evidence". I don't want the custom function to fail if the user is just trying to readd already added evidence.

philroyer-phantom replied:
Good call, that makes more sense.

philroyer-phantom commented on custom_functions/mark_evidence.py:
```diff
@@ -0,0 +1,66 @@
+    elif response.get('failed') and response.get('message'):
+        phantom.debug(f"{content_type} \'{container_id}\' {response['message']}")
+    else:
+        phantom.error(f"Unknown error in mark_evidence: {response}")
```
I would also raise a RuntimeError here, which is the way we are reporting failures in custom functions when needed.

kelby-shelton replied:
Done

philroyer-phantom reviewed (changes requested):
Just a few more small changes and then we should be good to merge. Thank you for your patience.

kelby-shelton committed:
Made recommended adjustments

Added changes for vault_id and extra error handling

philroyer-phantom reviewed (approved):
Looks good! Thank you for the back and forth.
